### PR TITLE
Addon-docs: MDX Linking

### PR DESCRIPTION
**Pull request by patricklafrance**
Issue: #8184

## What I did

Improved MDX linking:

- It's now possible to link to an in page anchor.
- URL to other stories are now automatically transformed to use the top level iframe url instead of the preview iframe url.
- It's now possible to link to an anchor of another story
- Calling a story with a hash will scroll to the matching element.
- Github style autolink headers are automatically generated
- target and rel attributes is automatically added to all external links (view [remark-external-link plugin](https://github.com/remarkjs/remark-external-links) for more info)

sb_linking

## How to test

- Is this testable with Jest or Chromatic screenshots? No
- Does this need a new example in the kitchen sink apps? Already done
- Does this need an update to the documentation? Maybe?

It can be tested with the `?path=/docs/addons-docs-markdown-docs--page` story of the official storybook.

### CODE REVIEW

**Comment by shilman** [comment]
@patricklafrance this is INCREDIBLE! 💯💯💯 

One small request is to fix the linking behavior. Storybook vs Github:

linking-sb

linking-gh

**Comment by patricklafrance** [comment]
@shilman done!

File: addons/docs/src/blocks/DocsContainer.tsx [hunk 1]
```diff
@@ -1,75 +1,66 @@
...
+    if (url.hash) {
+      const element = document.getElementById(url.hash.substring(1));
+      if (element) {
+        // Introducing a delay to ensure scrolling works when it's a full refresh.
+        setTimeout(() => {
+          scrollToElement(element);
+        }, 200);
+      }
+    } else {
+      let element = document.getElementById(anchorBlockIdFromId(storyId));
```
Comment (shilman) [inline addons/docs/src/blocks/DocsContainer.tsx]: element = getX || getY?

File: addons/docs/src/blocks/mdx.tsx [hunk 2]
```diff
@@ -1,14 +1,203 @@
-import * as React from 'react';
+import React, { FC, SyntheticEvent } from 'react';
+import { Source } from '@storybook/components';
+import { Code, components } from '@storybook/components/html';
```
Comment (shilman) [inline addons/docs/src/blocks/mdx.tsx]: { Code, a: A }

File: addons/docs/src/frameworks/common/preset.ts [hunk 3]
```diff
@@ -1,5 +1,9 @@
 /* eslint-disable import/no-extraneous-dependencies */
 import createCompiler from '@storybook/addon-docs/mdx-compiler-plugin';
+// @ts-ignore
```
Comment (shilman) [inline addons/docs/src/frameworks/common/preset.ts]: `declare module` instead?

**Review by shilman — APPROVED** [review]
THIS IS SO RAD

**Comment by fooontic** [comment]
Where can I find the sample code snippet of this feature?

**Comment by tylucaskelley** [comment]
Perhaps I'm missing something, but this doesn't appear to work as expected in the official docs - https://next--storybookjs.netlify.app/official-storybook/?path=/docs/addons-docs-markdown-docs--page

Clicking a heading just takes me to a blank page because it tries to go to the root path

**Comment by sag1v** [comment]
> Perhaps I'm missing something, but this doesn't appear to work as expected in the official docs - https://next--storybookjs.netlify.app/official-storybook/?path=/docs/addons-docs-markdown-docs--page
> 
> Clicking a heading just takes me to a blank page because it tries to go to the root path

there's an open issue with a PR #16918

**Pull request merged**